fix(android): improve error handling for external storage file access

ish1416 opened the pull request:
<!--
Before submitting a PR, please read https://github.com/tauri-apps/tauri/blob/dev/.github/CONTRIBUTING.md#pull-request-guidelines

1. Give the PR a descriptive title.

  Examples of good title:
    - fix(windows): fix race condition in event loop
    - docs: update example for `App::show`
    - feat: add `Window::set_fullscreen`

  Examples of bad title:
    - fix #7123
    - update docs
    - fix bugs

2. If there is a related issue, reference it in the PR text, e.g. closes #123.
3. If this change requires a new version, then add a change file in `.changes` directory with the appropriate bump, see https://github.com/tauri-apps/tauri/blob/dev/.changes/README.md
4. Ensure that all your commits are signed https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits
5. Ensure `cargo test` and `cargo clippy` passes.
6. Propose your changes as a draft PR if your work is still in progress.
-->

Fixes 500 error when accessing local video files in Android external storage directory via `convertFileSrc`. Added better error handling and logging for Android external storage access to help diagnose permission and accessibility issues.

Closes #14432

- Enhanced Android-specific error handling in asset protocol handler
- Added file existence checks before attempting to open files  
- Added detailed logging for debugging Android external storage access
- Improved error messages with specific guidance for storage permissions
- Added changelog entry in `.changes` directory

ish1416 committed:
fix(android): improve error handling for external storage file access

- Add better error handling and logging for Android external storage paths
- Provide more descriptive error messages when file access fails
- Check file existence and permissions before attempting to open files
- Helps diagnose issues with accessing files in /storage/emulated/0/Android/data/ paths

Fixes #14432

ish1416 commented:
@FabianLars Can you review this PR?

FabianLars commented on crates/tauri/src/protocol/asset.rs:
```diff
@@ -48,8 +48,45 @@ fn get_response(
+      if !std::path::Path::new(&path).exists() {
+        log::error!("File does not exist at path: {}", path);
+        return resp.status(404).body(Vec::new().into()).map_err(Into::into);
+      }
```
this doesn't have to be android-only. it should be valuable information on other platforms as well.

FabianLars commented on crates/tauri/src/protocol/asset.rs:
```diff
@@ -48,8 +48,45 @@ fn get_response(
+  #[cfg(target_os = "android")]
+  {
+    if path.starts_with("/storage/emulated/0/Android/data/") {
+      log::debug!("Attempting to access Android external storage path: {}", path);
```
i don't think this adds much value

FabianLars commented on crates/tauri/src/protocol/asset.rs:
```diff
@@ -48,8 +48,45 @@ fn get_response(
+          log::error!("Failed to get metadata for {}: {}", path, e);
+          return resp.status(500).body(format!("Failed to access file metadata: {}", e).into_bytes().into()).map_err(Into::into);
+        }
+      }
```
this also doesn't have to be android-only.

FabianLars commented on crates/tauri/src/protocol/asset.rs:
```diff
@@ -48,8 +48,45 @@ fn get_response(
+      // Check if we have read permissions
+      match std::fs::metadata(&path) {
+        Ok(metadata) => {
+          log::debug!("File metadata for {}: size={}, readonly={}", path, metadata.len(), metadata.permissions().readonly());
```
This log looks more like the verbosity that's hidden behind the `tracing` feature flag but we're _very_ inconsistent about this :/

FabianLars commented on crates/tauri/src/protocol/asset.rs:
```diff
@@ -48,8 +48,45 @@ fn get_response(
+        {
+          if path.starts_with("/storage/emulated/0/Android/data/") {
+            log::error!("Failed to open Android external storage file '{}': {}. This may be due to missing storage permissions or the file being inaccessible from the webview context.", path, e);
+            log::error!("Ensure your app has the necessary storage permissions and the file is in an accessible location.");
```
This second line can be derived from the first one so let's remove it to not be _that_ spammy.

FabianLars reviewed (commented):
thanks for the pr :) a few ideas since most of your code should offer value on other platforms as well, not just android.

ish1416 commented:
@FabianLars Thanks for the review . Ill use your suggestions!

ish1416 committed:
refactor: apply reviewer feedback

- Make file existence and metadata checks platform-agnostic (not Android-only)
- Remove verbose debug logging that doesn't add value
- Simplify error messages to reduce spam
- Keep Android-specific context only where necessary

ish1416 commented:
@FabianLars Ive changed in the latest commits whats have been asked can you review it now?

FabianLars commented on crates/tauri/src/protocol/asset.rs:
```diff
@@ -48,8 +48,31 @@ fn get_response(
+  // Check if we have read permissions
+  if let Err(e) = std::fs::metadata(&path) {
+    log::error!("Failed to get metadata for {}: {}", path, e);
+    return resp.status(500).body(format!("Failed to access file metadata: {}", e).into_bytes().into()).map_err(Into::into);
```
maybe we shouldn't use `500` here since that's our fallback on any kind of error. Any other ideas? 403 comes to mind but we already use that above 🤔

ish1416 replied:
Okay will sure fix this

ish1416 replied:
I used 422 instead of 403.... Its better i Guess

ish1416 committed:
fix: use 422 instead of 500 for metadata access failures

Use 422 Unprocessable Entity instead of 500 Internal Server Error
for metadata access failures, as it's more semantically correct.
The file exists and path is valid, but can't be processed due to
metadata access issues (permissions, locks, etc.).

FabianLars committed:
Merge branch 'dev' into fix/android-external-files-access

FabianLars committed:
handle errors when opening file
